File: patterns/ai/cost-and-tokens/files/py/src/ai/budget.py
"""Per-run cost accounting.

The gateway enforces a ceiling on a *single* call. That does not stop the
failure that actually costs money: an agent loop making 400 individually
reasonable calls. This tracks the run as a whole.

Uses ``contextvars`` so call sites do not have to thread a ledger through every
function — the ledger follows the execution context instead.
"""

from __future__ import annotations

import uuid
from contextlib import contextmanager
from contextvars import ContextVar
from dataclasses import dataclass, field
from typing import Iterator

from ..config.env import settings


@dataclass
class CallRecord:
    model: str
    label: str
    input_tokens: int
    output_tokens: int
    cached_input_tokens: int
    cost_usd: float
    ms: float


@dataclass
class RunLedger:
    run_id: str
    ceiling_usd: float
    calls: list[CallRecord] = field(default_factory=list)

    @property
    def total_usd(self) -> float:
        return sum(c.cost_usd for c in self.calls)


class RunBudgetExceededError(RuntimeError):
    pass


_current: ContextVar[RunLedger | None] = ContextVar("run_ledger", default=None)


def current_run() -> RunLedger | None:
    return _current.get()


@contextmanager
def run_budget(
    run_id: str | None = None, ceiling_usd: float | None = None
) -> Iterator[RunLedger]:
    """Run a block inside a budgeted context. Nested blocks share the ledger."""
    existing = _current.get()
    if existing is not None:
        yield existing
        return

    ledger = RunLedger(
        run_id=run_id or str(uuid.uuid4()),
        ceiling_usd=ceiling_usd if ceiling_usd is not None else settings.ai_max_usd_per_run,
    )
    token = _current.set(ledger)
    try:
        yield ledger
    finally:
        _current.reset(token)


def record(call: CallRecord) -> None:
    """Record a completed call. Raises once the run's total passes its ceiling."""
    ledger = _current.get()
    if ledger is None:
        return  # accounting is optional; not being in a run is fine
    ledger.calls.append(call)

    if ledger.total_usd > ledger.ceiling_usd:
        top = sorted(ledger.calls, key=lambda c: c.cost_usd, reverse=True)[:3]
        detail = ", ".join(f"{c.label} (${c.cost_usd:.4f})" for c in top)
        raise RunBudgetExceededError(
            f"Run {ledger.run_id} spent ${ledger.total_usd:.4f} over "
            f"{len(ledger.calls)} calls, past its ${ledger.ceiling_usd:.2f} "
            f"ceiling. Most expensive: {detail}"
        )


def summarize(ledger: RunLedger) -> str:
    """Human-readable summary.

    Print this at the end of every CLI run and job: cost you cannot see is cost
    you do not manage.
    """
    by_model: dict[str, dict[str, float]] = {}
    for c in ledger.calls:
        e = by_model.setdefault(c.model, {"n": 0, "usd": 0.0, "in": 0, "out": 0})
        e["n"] += 1
        e["usd"] += c.cost_usd
        e["in"] += c.input_tokens
        e["out"] += c.output_tokens

    lines = [
        f"run {ledger.run_id}: {len(ledger.calls)} calls, "
        f"${ledger.total_usd:.4f} of ${ledger.ceiling_usd:.2f}"
    ]
    for model, e in by_model.items():
        lines.append(
            f"  {model:<28} {int(e['n']):>3} calls  "
            f"{int(e['in'])}→{int(e['out'])} tok  ${e['usd']:.4f}"
        )

    total_in = sum(c.input_tokens for c in ledger.calls)
    cached = sum(c.cached_input_tokens for c in ledger.calls)
    if total_in:
        lines.append(f"  cache hit rate: {cached / total_in * 100:.0f}% of input tokens")
    return "\n".join(lines)

File: patterns/ai/cost-and-tokens/files/ts/src/ai/budget.ts
/**
 * Per-run cost accounting.
 *
 * The gateway enforces a ceiling on a *single* call. That does not stop the
 * failure that actually costs money: an agent loop making 400 individually
 * reasonable calls. This tracks the run as a whole.
 *
 * Uses AsyncLocalStorage so call sites do not have to thread a ledger through
 * every function — the ledger follows the async context instead.
 */
import { AsyncLocalStorage } from "node:async_hooks";
import { env } from "../config/env.js";

export interface CallRecord {
  model: string;
  label: string;
  inputTokens: number;
  outputTokens: number;
  cachedInputTokens: number;
  costUsd: number;
  ms: number;
}

export interface RunLedger {
  runId: string;
  ceilingUsd: number;
  calls: CallRecord[];
}

export class RunBudgetExceededError extends Error {}

const storage = new AsyncLocalStorage<RunLedger>();

export function currentRun(): RunLedger | undefined {
  return storage.getStore();
}

/** Run `fn` inside a budgeted context. Nested calls share the same ledger. */
export async function withRunBudget<T>(
  opts: { runId?: string; ceilingUsd?: number },
  fn: (ledger: RunLedger) => Promise<T>,
): Promise<T> {
  const ledger: RunLedger = {
    runId: opts.runId ?? crypto.randomUUID(),
    ceilingUsd: opts.ceilingUsd ?? env.AI_MAX_USD_PER_RUN,
    calls: [],
  };
  return storage.run(ledger, () => fn(ledger));
}

/** Record a completed call. Throws once the run's total passes its ceiling. */
export function record(call: CallRecord): void {
  const ledger = storage.getStore();
  if (!ledger) return; // accounting is optional; not being in a run is fine
  ledger.calls.push(call);

  const total = totalUsd(ledger);
  if (total > ledger.ceilingUsd) {
    throw new RunBudgetExceededError(
      `Run ${ledger.runId} spent $${total.toFixed(4)} over ${ledger.calls.length} calls, ` +
        `past its $${ledger.ceilingUsd.toFixed(2)} ceiling. ` +
        `Most expensive: ${topCalls(ledger, 3).map((c) => `${c.label} ($${c.costUsd.toFixed(4)})`).join(", ")}`,
    );
  }
}

export const totalUsd = (l: RunLedger): number => l.calls.reduce((n, c) => n + c.costUsd, 0);

const topCalls = (l: RunLedger, n: number): CallRecord[] =>
  [...l.calls].sort((a, b) => b.costUsd - a.costUsd).slice(0, n);

/**
 * Human-readable summary. Print this at the end of every CLI run and job:
 * cost you cannot see is cost you do not manage.
 */
export function summarize(l: RunLedger): string {
  const byModel = new Map<string, { n: number; usd: number; in: number; out: number }>();
  for (const c of l.calls) {
    const e = byModel.get(c.model) ?? { n: 0, usd: 0, in: 0, out: 0 };
    byModel.set(c.model, {
      n: e.n + 1,
      usd: e.usd + c.costUsd,
      in: e.in + c.inputTokens,
      out: e.out + c.outputTokens,
    });
  }

  const cached = l.calls.reduce((n, c) => n + c.cachedInputTokens, 0);
  const input = l.calls.reduce((n, c) => n + c.inputTokens, 0);
  const lines = [
    `run ${l.runId}: ${l.calls.length} calls, $${totalUsd(l).toFixed(4)} of $${l.ceilingUsd.toFixed(2)}`,
  ];
  for (const [model, e] of byModel) {
    lines.push(`  ${model.padEnd(28)} ${String(e.n).padStart(3)} calls  ${e.in}→${e.out} tok  $${e.usd.toFixed(4)}`);
  }
  if (input > 0) {
    lines.push(`  cache hit rate: ${((cached / input) * 100).toFixed(0)}% of input tokens`);
  }
  return lines.join("\n");
}
